### DISCUSSION TITLE
Sveltekit form Actions throws cross-site (cors) error - but only with Vercel

### DISCUSSION THREAD

**Discussion by DrMandible** [vercel/community#3390] [Help]
### Summary

I'm using Sveltekit (Vercel adapter) and using the default form actions example in the Sveltekit docs. But I get a cors error when I submit any form action. 

When I run this app using "vercel dev" I get the following error: "Cross-site POST form submissions are forbidden". When I run this exact same code with "npm run dev", it performs as expected.

I even tested disabling the csrf check. But that did not solve the problem either.

### Steps to Reproduce

The following code is from path: "src/routes" but I have also tested various other routes with the same results.

+page.svelte:
```
<form method="POST">
    <label>
        Email
        <input name="email" type="email" />
    </label>
    <label>
        Password
        <input name="password" type="password" />
    </label>
    <button>Log in</button>
</form>
```

+page.server.js
```
export const actions = {
    default: async (event) => {
        console.log('hello!?')
        // TODO log the user in
    }
};
```

And here is the package and config files. Nothing out of the ordinary.

package.json
```
{
	"name": "test",
	"version": "0.4.0",
	"private": true,
	"engines": {
		"node": ">=18"
	},
	"scripts": {
		"dev": "vite dev",
		"build": "vite build",
		"preview": "vite preview",
		"lint": "prettier --plugin-search-dir . --check . && eslint .",
		"format": "prettier --plugin-search-dir . --write ."
	},
	"devDependencies": {
		"@sveltejs/adapter-auto": "^2.1.0",
		"@sveltejs/kit": "^1.22.3",
		"autoprefixer": "^10.4.14",
		"eslint": "^8.28.0",
		"eslint-config-prettier": "^8.5.0",
		"eslint-plugin-svelte": "^2.30.0",
		"postcss": "^8.4.24",
		"postcss-load-config": "^4.0.1",
		"prettier": "^2.8.0",
		"prettier-plugin-svelte": "^2.10.1",
		"svelte": "^4.0.5",
		"tailwindcss": "^3.3.2",
		"vite": "^4.4.2"
	},
	"type": "module",
	"dependencies": {
		"@popperjs/core": "^2.11.8",
		"@sveltejs/adapter-vercel": "^3.0.2",
		"@vercel/postgres": "^0.4.1",
		"flowbite": "^1.7.0",
		"flowbite-svelte": "^0.39.2",
		"latlon-geohash": "^2.0.0",
		"leaflet": "^1.9.4",
		"luxon": "^3.3.0",
		"node": "^18.17.0",
		"tailwind-merge": "^1.13.2",
		"vercel": "^24.2.4"
	}
}
```

svelte.config.js
```
import { vitePreprocess } from '@sveltejs/kit/vite';
import adapter from '@sveltejs/adapter-vercel';

/** @type {import('@sveltejs/kit').Config} */
const config = {
	kit: {
		adapter: adapter()
	},
	preprocess: [vitePreprocess({})]
};

export default config;
```

vite.config.js
```
import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vite';

export default defineConfig({
	plugins: [sveltekit()]
});
```

**Comment by amyegan** [comment]
Hey @DrMandible. I have not been able to observe this problem with any of my SvelteKit apps. Would you be able to put together a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) in a public repo? That would give me a better view into what is going wrong

**Reply by DrMandible** [reply]
I made a public repo by forking the Vercel Sveltekit boilerplate found here: https://vercel.com/templates/next.js/sveltekit-boilerplate

I made no changes other than to replace the html in +page.svelte with the simplest example form from the Sveltekit docs here: https://kit.svelte.dev/docs/form-actions

The cors error is still occurring. Here is the link to my public repo where you can reproduce the error: https://github.com/DrMandible/lls

Edit: I just noticed in the description for the Vercel boilerplate, it recommends running the app with "npm run dev". But I thought the preferred way to test an app for Vercel deployment is to use "vercel dev". Should I be using "npm run dev" instead? If so, what is the purpose of "vercel dev"?

**Reply by Arro** [reply]
@DrMandible  did you end up figuring this out?  I'm having this issue presently.

**Reply by ghost** [reply]
Same, we have the same issue, what we do is that we set checkOrigin to false when there is no VERCEL_ENV defined. Maybe this comes from the adapter-vercel going nuts with ORIGIN on localhost.

**Reply by 16Integer** [reply]
How did you check if you were running in production or development to set checkOrigin?

**Reply by dominicfarr** [reply]
any working solution for this?

**Reply by Arro** [reply]
I just use `npm run dev`

**Reply by amyegan** [reply]
I was able to build a repro example app and do a bit of testing with it. It seems to be a difference in port numbers causing the CORS error. 

The easiest way I found to get it working is if I go to the project's [Build and Development Settings](https://vercel.com/docs/deployments/configure-a-build#build-and-development-settings) and override the [Development Command](https://vercel.com/docs/deployments/configure-a-build#development-command). Either `vite` or `vite --port 3000` work reliably for me.

<img width="727" alt="" src="https://github.com/vercel/community/assets/8247105/ab10f115-2954-43e7-a732-587e0807a241">

I'll work with the config team to update the default command.

**Comment by dominicfarr** [comment]
I'm having a similar issue.  
`vite dev` works for GET requests and form POST

`vercel dev` doesn't work

Form POST triggers this cors error. 
`Cross-site POST form submissions are forbidden`

I've added headers to allow cors in the vercel.json 
```json
{
  "buildCommand": "pnpm turbo build",
  "ignoreCommand": "pnpm dlx turbo-ignore",
  "headers": [
    {
      "source": "/(.*)",
      "headers": [
        {
          "key": "Access-Control-Allow-Origin",
          "value": "*"
        },
        {
          "key": "Access-Control-Allow-Methods",
          "value": "GET, POST, PUT, PATCH, DELETE, OPTIONS"
        },
        {
          "key": "Access-Control-Allow-Headers",
          "value": "Content-Type, Authorization"
        }
      ]
    }
  ]
}

```
I have hooks.server.js that logs some information and response to OPTION requests. 
```js
export async function handle({ event, resolve }) {
  const { request } = event;

  if (request.method === "OPTIONS") {
    return new Response(null, {
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods":
          "GET, POST, PUT, PATCH, DELETE, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type, Authorization",
      },
    });
  }

  // If it's not a preflight request, continue to the next middleware or endpoint
  const response = await resolve(event);
  return response;
}
```

What an I missing here?

**Reply by dominicfarr** [reply]
I did find this in the docs 
https://kit.svelte.dev/docs/adapter-node#environment-variables-origin-protocolheader-hostheader-and-port-header

```
If adapter-node can't correctly determine the URL of your deployment, you may experience this error when using [form actions](https://kit.svelte.dev/docs/form-actions):

Cross-site POST form submissions are forbidden
```

**Reply by dominicfarr** [reply]
adding `export ORIGIN=http://localhost:3000` and then running `vercel dev` doesn't improve this situation sadly.

**Reply by dominicfarr** [reply]
This is my current workaround for `vercel dev` to work. 

```js
import adapter from "@sveltejs/adapter-vercel";
import { vitePreprocess } from "@sveltejs/vite-plugin-svelte";

/** @type {import('@sveltejs/kit').Config} */
const config = {
  kit: {
    adapter: adapter(),
    csrf: {
      checkOrigin: process.env.NODE_ENV === "development" ? false : true,
    },
  },
  preprocess: vitePreprocess(),
};

export default config;

```

**Reply by DrMandible** [reply]
I had to use the hack @dominicfarr used while I developed and waited for Vercel to answer my question. Since they did not, perhaps could not, **I ultimately abandoned Vercel.** 

My project was intended for an enterprise environment for a client with demanding privacy needs. I can't be opening their code up to these sorts of vulnerabilities.

**Comment by AshleyJackson** [comment]
I'm not overly sure if there is any correlation here, however I was experiencing this issue in both Dev (behind NGROK) and in production (Cloudflare)

I found that the "checkOrigin" worked in Dev, but not in production. I later was able to resolve this in production by disabling Cloudflare's SSL Proxy, and just have the DNS record point directly to my production environment (Vercel).
